fl4health/server/tabular_feature_alignment_server.py

```diff
@@ -0,0 +1,143 @@
+import random
+import warnings
+from functools import partial
+from logging import INFO
+from typing import Callable, List, Optional, Tuple
+
+from flwr.common.logger import log
+from flwr.common.typing import Config
+from flwr.server.client_manager import ClientManager
+from flwr.server.history import History
+
+from fl4health.checkpointing.checkpointer import TorchCheckpointer
+from fl4health.feature_alignment.constants import (
+    CURRENT_SERVER_ROUND,
+    FEATURE_INFO,
+    FORMAT_SPECIFIED,
+    INPUT_DIMENSION,
+    OUTPUT_DIMENSION,
+)
+from fl4health.feature_alignment.tab_features_info_encoder import TabularFeaturesInfoEncoder
+from fl4health.reporting.fl_wanb import ServerWandBReporter
+from fl4health.server.base_server import FlServer
+from fl4health.server.polling import poll_clients
+from fl4health.strategies.basic_fedavg import BasicFedAvg
+
+
+class TabularFeatureAlignmentServer(FlServer):
+    """
+    This server is used when the clients all have tabular data that needs to be
+    aligned.
+
+    tab_features_info: the information that is required for aligning client features.
+    If it is not specified, then the server will randomly poll a client and gather
+    this information from its data source.
+    """
+
+    def __init__(
+        self,
+        client_manager: ClientManager,
+        config: Config,
+        initialize_parameters: Callable,
+        strategy: BasicFedAvg,
+        wandb_reporter: Optional[ServerWandBReporter] = None,
+        checkpointer: Optional[TorchCheckpointer] = None,
+        tab_features_info: Optional[TabularFeaturesInfoEncoder] = None,
+    ) -> None:
+        assert isinstance(strategy, BasicFedAvg)
+        if strategy.on_fit_config_fn is not None:
+            warnings.warn("strategy.on_fit_config_fn will be overwritten.")
+        if strategy.initial_parameters is not None:
+            warnings.warn("strategy.initial_parameters will be overwritten.")
+
+        super().__init__(client_manager, strategy, wandb_reporter, checkpointer)
+        # The server performs one or two rounds of polls before the normal federated training.
+        # The first one gathers feature information if the server does not already have it,
+        # and the second one gathers the input/output dimensions of the model.
+        self.initial_polls_complete = False
+        self.tab_features_info = tab_features_info
+        self.config = config
+        self.initialize_parameters = initialize_parameters
+        self.format_info_gathered = False
+        # casting self.strategy to BasicFedAvg so its on_fit_config_fn can be specified.
+        self.strategy: BasicFedAvg
+        self.strategy.on_fit_config_fn = partial(fit_config, self.config, self.format_info_gathered)
+
+    def fit(self, num_rounds: int, timeout: Optional[float]) -> History:
+        """Run federated averaging for a number of rounds."""
+
+        assert isinstance(self.strategy, BasicFedAvg)
```

Comment:
I would recommend we test whether the `initial_parameters` of the strategy have been set here. If they have, we should warn the user that they will be ignored in favor of the `self.initialize_parameters` function